docs/advanced/pycpp/object.rst

```diff
@@ -62,6 +64,40 @@ This example obtains a reference to the Python ``Decimal`` class.
     py::object scipy = py::module::import("scipy");
     return scipy.attr("__version__");
 
+.. _implicit_casting:
+
+Implicit (automatic) casting
+============================
+
+Instead of using a generic :class:`object` as return type, it is possible to
+specialize to a subtype, e.g., to use the lookup syntax of :class:`dict`.
+
+.. note::
+    Casting to the wrong type will currently lead to downstream errors, e.g.,
+    when casting a Python dict to :class:`list` and using the ``append`` method.
+    In the future, such a bad cast will likely result in a :class:`cast_error`,
+    see `PR #2349 <https://github.com/pybind/pybind11/pull/2349>`_.
```

Comment:
Maybe this isn't necessary to add? I'm a bit hesitant to add open issues/PRs to this part of the docs. I'd say it's up to #2349 to add something here.

Reply:
But I guess it's fine as well. As you please.

(Feel free to ping #2349, so we won't forget there.)

Reply:
I turned the PR part into an RST comment.

On the larger context: I turns out (new to me) that `py::list lst = obj;` first tries to move-construct `lst` from `obj`, and if that's not possible, does something like `list(obj)` would do in Python. So the old claim that casting a `dict` to `py::list` would fail later on was wrong: it would create a list of the dict keys.

 I've adapted the note box accordingly. Please have a look.

Reply:
Ha, it does? OK, then. This is also related to #1201, so with a bit of luck, we can straighten out things a bit more in the future anyway (conversions vs. moving vs. reinterpreting, etc.) !